### PR TITLE
fix: rpc getblock and getblockstats for blocks with withdrawal transactions (asset unlock)

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
https://github.com/dashpay/dash/issues/6335

## What was done?
Applied fixes for `getblock` rpc and `getblockstats` rpc to make them work with withdrawal transactions (asset unlock).

## How Has This Been Tested?
Run updated functional test `feature_asset_locks.py` without the fix causes a failure:
```
2024-10-22T12:01:35.902000Z TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/home/knst/projects/dash-reviews/test/functional/test_framework/test_framework.py", line 160, in main
    self.run_test()
  File "/home/knst/projects/dash-reviews/test/functional/feature_asset_locks.py", line 273, in run_test
    self.test_asset_unlocks(node_wallet, node, pubkey)
  File "/home/knst/projects/dash-reviews/test/functional/feature_asset_locks.py", line 410, in test_asset_unlocks
    self.log.info(f"block info: {node.getblock(block_asset_unlock, 2)}")
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/knst/projects/dash-reviews/test/functional/test_framework/coverage.py", line 49, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/knst/projects/dash-reviews/test/functional/test_framework/authproxy.py", line 148, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: Internal bug detected: "MoneyRange(fee)"
core_write.cpp:338 (TxToUniv)
Please report this issue here: https://github.com/dashpay/dash/issues
 (-1)
```

With patch functional test `feature_asset_locks.py` succeed.

## Breaking Changes
N/A

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by ogabrielides — DISMISSED** [review]
LGTM

**Review by kwvg — DISMISSED** [review]
ACK 6cd74906edab2e9462d37ae5fe20f150cdfd0e82

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -2450,6 +2451,15 @@ static RPCHelpMan getblockstats()
             }
 
             CAmount txfee = tx_total_in - tx_total_out;
+
+            if (tx->nType == TRANSACTION_ASSET_UNLOCK) {
+                if (const auto opt_assetUnlockTx = GetTxPayload<CAssetUnlockPayload>(*tx)) {
+                    txfee = opt_assetUnlockTx->getFee();
+                } else {
+                    txfee = 0;
+                }
```
Comment (PastaPastaPasta) [inline src/rpc/blockchain.cpp:2460]:
```suggestion
                const auto opt_assetUnlockTx = GetTxPayload<CAssetUnlockPayload>(*tx);
                CHECK_NONFATAL(opt_assetUnlockTx.has_value());
                txfee = opt_assetUnlockTx->getFee();
```

Comment (UdjinM6) [reply src/rpc/blockchain.cpp:2460]: Taking a step further... How about bc3a811730?

Comment (PastaPastaPasta) [reply src/rpc/blockchain.cpp:2460]: Looks reasonable to me!

Comment (knst) [reply src/rpc/blockchain.cpp:2460]:
```
    if (calculate_fee) {
        CAmount fee = amt_total_in - amt_total_out;
        if (tx.nType == TRANSACTION_ASSET_UNLOCK) {
            fee = CHECK_NONFATAL(GetTxPayload<CAssetUnlockPayload>(tx))->getFee();
        }
```

Need to call `GetTxPayload` twice, I guess over-head is not huge?

Comment (UdjinM6) [reply src/rpc/blockchain.cpp:2460]: That's just reading/deserializing from datastream, no heavy ops. Should be fine imo.

**Comment by UdjinM6** [comment]
bb6d386cc92a68746ce4ef57555951fffef33eb7:
```diff
diff --git a/src/rpc/masternode.cpp b/src/rpc/masternode.cpp
index 95487525e1..26604bc4c9 100644
--- a/src/rpc/masternode.cpp
+++ b/src/rpc/masternode.cpp
@@ -407,7 +407,9 @@ static RPCHelpMan masternode_payments()
             if (tx->IsCoinBase()) {
                 continue;
             }
-            if (!tx->IsPlatformTransfer()) {
+            if (tx->IsPlatformTransfer()) {
+                nBlockFees += CHECK_NONFATAL(GetTxPayload<CAssetUnlockPayload>(*tx))->getFee();
+            } else {
                 CAmount nValueIn{0};
                 for (const auto& txin : tx->vin) {
                     uint256 blockHashTmp;
@@ -415,10 +417,6 @@ static RPCHelpMan masternode_payments()
                     nValueIn += txPrev->vout[txin.prevout.n].nValue;
                 }
                 nBlockFees += nValueIn - tx->GetValueOut();
-            } else {
-                if (const auto opt_assetUnlockTx = GetTxPayload<CAssetUnlockPayload>(*tx)) {
-                    nBlockFees += opt_assetUnlockTx->getFee();
-                }
             }
         }
 
```

**Comment by knst** [comment]
```
-            if (!tx->IsPlatformTransfer()) {
+            if (tx->IsPlatformTransfer()) {
```

why? I submitted fix... though it requires new CI run and re-test that there's no silly bug introduced...

**Comment by UdjinM6** [comment]
no negative condition + `IsPlatformTransfer` and platform related fee logic are close to each other

**Comment by UdjinM6** [comment]
f6169fade4303284aa97aba7f3869c761e0ab242 even better 👍

**Review by kwvg — APPROVED** [review]
ACK f6169fade4303284aa97aba7f3869c761e0ab242

File: src/rpc/masternode.cpp [hunk 2]
```diff
@@ -406,6 +407,11 @@ static RPCHelpMan masternode_payments()
             if (tx->IsCoinBase()) {
                 continue;
             }
+            if (tx->IsPlatformTransfer()) {
```
Comment (PastaPastaPasta) [inline src/rpc/masternode.cpp:410]: we don't have IsPlatformTransfer in v21.1.x

Comment (knst) [reply src/rpc/masternode.cpp:410]: Let's backport this function too, it's a single-liner

**Review by PastaPastaPasta — APPROVED** [review]
utACK b9a46f6d2c14a07803cc26cf79b3a0d97fe8e486

**Review by kwvg — APPROVED** [review]
utACK b9a46f6d2c14a07803cc26cf79b3a0d97fe8e486

File: src/rpc/masternode.cpp (hunk 2 above)
Comment (UdjinM6) [reply src/rpc/masternode.cpp:410]: Agree, looks like #6131 can be backported with no merge conflicts. Compiles with no issues too.

**Review by UdjinM6 — APPROVED** [review]
utACK b9a46f6d2c14a07803cc26cf79b3a0d97fe8e486

**Review by ogabrielides — APPROVED** [review]
utACK https://github.com/dashpay/dash/commit/b9a46f6d2c14a07803cc26cf79b3a0d97fe8e486

**Pull request merged**